File: sdk-go/dephealth/checks/postgres.go
package checks

import (
	"context"
	"database/sql"
	"fmt"
	"net"
	"strings"

	_ "github.com/jackc/pgx/v5/stdlib" // PostgreSQL driver

	"github.com/BigKAA/topologymetrics/sdk-go/dephealth"
)

// PostgresOption configures the PostgresChecker.
type PostgresOption func(*PostgresChecker)

// PostgresChecker performs health checks against a PostgreSQL database.
// Supports two modes:
//   - Standalone: creates a new connection per check using DSN built from endpoint
//   - Pool: uses an existing *sql.DB connection pool
type PostgresChecker struct {
	db    *sql.DB // nil = standalone, non-nil = pool mode
	dsn   string  // custom DSN for standalone mode (overrides endpoint-based DSN)
	query string  // health check query
}

// WithPostgresDB sets an existing connection pool for pool mode.
func WithPostgresDB(db *sql.DB) PostgresOption {
	return func(c *PostgresChecker) {
		c.db = db
	}
}

// WithPostgresDSN sets a custom DSN for standalone mode.
// If set, the endpoint host/port are ignored.
func WithPostgresDSN(dsn string) PostgresOption {
	return func(c *PostgresChecker) {
		c.dsn = dsn
	}
}

// WithPostgresQuery sets the health check SQL query (default "SELECT 1").
func WithPostgresQuery(query string) PostgresOption {
	return func(c *PostgresChecker) {
		c.query = query
	}
}

// NewPostgresChecker creates a new PostgreSQL health checker with the given options.
func NewPostgresChecker(opts ...PostgresOption) *PostgresChecker {
	c := &PostgresChecker{
		query: "SELECT 1",
	}
	for _, opt := range opts {
		opt(c)
	}
	return c
}

// Check performs a health check against the PostgreSQL endpoint.
// In pool mode, uses the existing *sql.DB. In standalone mode, opens a new connection.
func (c *PostgresChecker) Check(ctx context.Context, endpoint dephealth.Endpoint) error {
	if c.db != nil {
		return c.checkPool(ctx)
	}
	return c.checkStandalone(ctx, endpoint)
}

func (c *PostgresChecker) checkPool(ctx context.Context) error {
	rows, err := c.db.QueryContext(ctx, c.query)
	if err != nil {
		return classifyPostgresError(err, "pool")
	}
	return rows.Close()
}

func (c *PostgresChecker) checkStandalone(ctx context.Context, endpoint dephealth.Endpoint) error {
	dsn := c.dsn
	if dsn == "" {
		dsn = fmt.Sprintf("postgres://%s/postgres", net.JoinHostPort(endpoint.Host, endpoint.Port))
	}

	db, err := sql.Open("pgx", dsn)
	if err != nil {
		return fmt.Errorf("postgres open %s: %w", endpoint.Host, err)
	}
	defer func() { _ = db.Close() }()

	rows, err := db.QueryContext(ctx, c.query)
	if err != nil {
		return classifyPostgresError(err, endpoint.Host)
	}
	return rows.Close()
}

// classifyPostgresError wraps PostgreSQL errors with appropriate classification.
// Detects auth errors via SQLSTATE codes 28000/28P01 in the error message.
func classifyPostgresError(err error, target string) error {
	msg := err.Error()
	if strings.Contains(msg, "28000") || strings.Contains(msg, "28P01") ||
		strings.Contains(msg, "password authentication failed") {
		return &dephealth.ClassifiedCheckError{
			Category: dephealth.StatusAuthError,
			Detail:   "auth_error",
			Cause:    fmt.Errorf("postgres %s: %w", target, err),
		}
	}
	return fmt.Errorf("postgres query %s: %w", target, err)
}

// Type returns the dependency type for this checker.
func (c *PostgresChecker) Type() string {
	return string(dephealth.TypePostgres)
}

File: sdk-python/dephealth/checks/postgres.py
"""PostgreSQL health checker."""

from __future__ import annotations

from typing import Any

from dephealth.checker import CheckAuthError, CheckConnectionRefusedError, CheckTimeoutError
from dephealth.dependency import Endpoint


class PostgresChecker:
    """Health check for PostgreSQL via SELECT 1."""

    def __init__(
        self,
        timeout: float = 5.0,
        query: str = "SELECT 1",
        pool: Any = None,  # noqa: ANN401
        dsn: str = "",
    ) -> None:
        self._timeout = timeout
        self._query = query
        self._pool = pool
        self._dsn = dsn

    async def check(self, endpoint: Endpoint) -> None:
        """Execute SELECT 1 on PostgreSQL."""
        try:
            import asyncpg
        except ImportError:
            msg = "asyncpg is required for Postgres checker"
            raise ImportError(msg) from None

        if self._pool is not None:
            # Pool mode: use the existing connection pool.
            async with self._pool.acquire() as conn:
                await conn.fetchval(self._query)
            return

        # Standalone mode: new connection.
        dsn = self._dsn or f"postgresql://{endpoint.host}:{endpoint.port}"
        try:
            conn = await asyncpg.connect(dsn, timeout=self._timeout)
            try:
                await conn.fetchval(self._query)
            finally:
                await conn.close()
        except TimeoutError as exc:
            msg = f"Postgres connection to {endpoint.host}:{endpoint.port} timed out"
            raise CheckTimeoutError(msg) from exc
        except OSError as e:
            msg = f"Postgres connection to {endpoint.host}:{endpoint.port} refused: {e}"
            raise CheckConnectionRefusedError(msg) from e
        except Exception as e:
            _classify_postgres_error(e, endpoint)
            raise

    def checker_type(self) -> str:
        """Return the checker type."""
        return "postgres"


def _classify_postgres_error(err: Exception, endpoint: Endpoint) -> None:
    """Re-raise PostgreSQL auth errors as CheckAuthError."""
    msg = str(err)
    if "28000" in msg or "28P01" in msg or "password authentication failed" in msg:
        raise CheckAuthError(
            f"Postgres auth error at {endpoint.host}:{endpoint.port}: {msg}"
        ) from err
